### PR TITLE
Tier-3 batch 3: isqrt, roots, and integer logs for HeaplessBigInt

### PR DESCRIPTION
**Pull request by kaidokert**
Stacked on #159. Adds the integer-root and integer-log surface to `HeaplessBigInt`, mirroring the `FixedUInt` impls. All Nct-only — the bit search, Newton iteration, and divide-down loops all branch on value.

- `Isqrt` (const-num-traits) — bit-by-bit floor square root, plus the inherent `checked_isqrt` shim (unsigned isqrt never fails).
- `num_integer::Roots` (feature-gated) — `nth_root`, giving `sqrt`/`cbrt` for free.
- `Ilog2` / `Ilog10` / `Ilog` (const-num-traits) — `ilog*` + `checked_ilog*`.

### Width preservation
`Isqrt` and `Roots` seed their estimate at the operand width (`one` widened before the `<< bits`), so every intermediate `candidate * candidate` / `pow` / `/` resolves at the value width, bit for bit with the same-width `FixedUInt`. Pinned by heapless-internal `.len` tests. The `Ilog` family returns a `u32` count, so there's no result width to preserve.

Generic parity: `carrier_generic.rs::isqrt_and_ilogs` (ungated) and `carrier_num_traits.rs::roots_sqrt_cbrt_nth` (gated). Module width-contract table updated.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Add integer square root and integer logarithm support to HeaplessBigInt, along with num-integer Roots integration, while preserving operand width semantics and expanding test coverage.

New Features:
- Provide const-num-traits Isqrt implementation and inherent checked_isqrt for HeaplessBigInt
- Implement const-num-traits Ilog2, Ilog10, and Ilog for HeaplessBigInt, returning u32 counts
- Add num_integer::Roots (nth_root, sqrt, cbrt) support for HeaplessBigInt under the num-traits feature

Enhancements:
- Document HeaplessBigInt width behavior for Isqrt, nth_root, and integer log operations in the module contract table

Tests:
- Extend generic carrier tests to cover Isqrt and integer log behavior consistently across carriers
- Add num-traits-based carrier tests for Roots methods (sqrt, cbrt, nth_root) and their correctness over ranges
- Introduce focused unit tests for HeaplessBigInt isqrt, nth_root, and ilog implementations, including width preservation and edge cases

## Summary by CodeRabbit

* **New Features**
  * Added integer logarithm support for heapless big integers, including base 2, base 10, and arbitrary bases.
  * Added integer square-root support for heapless big integers.
  * Added square, cube, and nth-root operations for heapless big integers.
  * Added consistent logarithm, square-root, and root behavior across supported integer types.

* **Documentation**
  * Documented result-width behavior for root and logarithm operations.

* **Tests**
  * Added coverage for logarithms, square roots, and nth roots across supported numeric types.

### CODE REVIEW

File: src/heapless/roots_impl.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+        // Clamp to r^n <= self < (r+1)^n.
+        while x.pow(n) > *self {
+            x -= one_w;
+        }
+        let mut x_plus_one = x + one_w;
+        while x_plus_one.pow(n) <= *self {
+            x += one_w;
+            x_plus_one = x + one_w;
+        }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/roots_impl.rs:76]: @gemini-code-assist Fixed in 057b12a — all power evaluations in the Newton and clamp loops now go through checked_pow, treating overflow as arbitrarily large (quotient 0 / probe fails). Added a regression test: sqrt(u32::MAX) at width 1, which panicked before.

File: src/heapless/roots_impl.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        // Root of a value with fewer than `n` bits is 1.
+        if n as usize > bit_len {
+            return <Self as One>::one().widened(width);
+        }
+
+        let one_w = <Self as One>::one().widened(width);
+        let initial_exp = (bit_len as u32).div_ceil(n).max(1);
+        let mut x = one_w << (initial_exp as usize);
+
+        let n_val: Self = From::from(n);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/roots_impl.rs]: @gemini-code-assist Fixed in 057b12a — n and n-1 are now built value-based at the operand width via FromPrimitive::from_u64 (natural width, never panics) then widened, rather than the 4-byte u32 From. This also fixes the CAP < 4/word panic.

File: src/heapless/roots_impl.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        let bit_len = self.bit_length();
+        // Root of a value with fewer than `n` bits is 1.
+        if n as usize > bit_len {
+            return <Self as One>::one().widened(width);
+        }
+
+        let one_w = <Self as One>::one().widened(width);
+        let initial_exp = (bit_len as u32).div_ceil(n).max(1);
+        let mut x = one_w << (initial_exp as usize);
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/roots_impl.rs:43]: Fixed in 057b12a — the Newton constants are built at the operand width via FromPrimitive::from_u64. Regression test added for HeaplessBigInt<u8,1>::sqrt, which panicked on the old 4-byte From.

File: src/heapless/roots_impl.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            }
+            x = x_new;
+        }
+
+        // Clamp to r^n <= self < (r+1)^n.
+        while x.pow(n) > *self {
+            x -= one_w;
+        }
+        let mut x_plus_one = x + one_w;
+        while x_plus_one.pow(n) <= *self {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/roots_impl.rs]: Fixed in 057b12a — the (x+1)^n probe now uses checked_pow; None (overflow) reads as > self, so sqrt(u32::MAX) at width 1 returns 65535 instead of panicking. Covered by a regression test.

**Comment by kaidokert** [comment]
@sourcery-ai review

File: src/heapless/isqrt.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        let start_bit = self.bit_length().div_ceil(2);
+        let mut bit_pos = start_bit;
+        while bit_pos > 0 {
+            bit_pos -= 1;
+            let candidate = result | (one_w << bit_pos);
+            if candidate * candidate <= self {
+                result = candidate;
+            }
+        }
+        result
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/isqrt.rs]: @gemini-code-assist Done in a680b7f — switched to the digit-by-digit O(width²) method (add/sub/shift only) on both carriers to keep them in lockstep. On heapless the >>1/>>2 are sub-word shifts so len is preserved without an extra widen. The zero early-return was already width-correct: returning self keeps the operand width (covered by isqrt_preserves_width).

File: src/heapless/ilog.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            return None;
+        }
+        let mut n = self;
+        let mut count = 0u32;
+        while n >= base {
+            n /= base;
+            count += 1;
+        }
+        Some(count)
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/ilog.rs:93]: @gemini-code-assist Done in a680b7f — base 2 delegates to ilog2 and base 10 to ilog10, on both FixedUInt and HeaplessBigInt (kept aligned per the parity contract).
